File: countSubarrays.js
/*
You are given an integer array nums and a positive integer k.

Return the number of subarrays where the maximum element of nums appears at least k times in that subarray.

A subarray is a contiguous sequence of elements within an array.
*/
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var countSubarrays = function(nums, k) {
  let count = 0;
  let max_num = Math.max(...nums);
  let windowStart = 0;
  let maxCount = 0;

  for (let windowEnd = 0; windowEnd < nums.length; windowEnd++) {
    if (nums[windowEnd] === max_num) {
      maxCount++;
    }

    while (maxCount >= k) {
      count += nums.length - windowEnd;
      if (nums[windowStart] === max_num) {
        maxCount--;
      }
      windowStart++;
    }
  }

  return count;
};

/*
Example 1:
nums = [1,3,2,3,3], k=2, Expected output: 6
Example 2:
nums = [1,4,2,1], k=3, Expected output: 0
Example 3:
nums = [28,5,58,91,24,91,53,9,48,85,16,70,91,91,47,91,61,4,54,61,49], k=1, Expected output: 187
*/

File: countSubarrays.rb
=begin
You are given an integer array nums and a positive integer k.

Return the number of subarrays where the maximum element of nums appears at least k times in that subarray.

A subarray is a contiguous sequence of elements within an array.
=end
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def count_subarrays(nums, k)
  count = 0
  max_num = nums.max
  window_start = 0
  max_count = 0

  (0...nums.length).each do |window_end|
    if nums[window_end] == max_num
      max_count += 1
    end

    while max_count >= k
      count += nums.length - window_end
      if nums[window_start] == max_num
        max_count -= 1
      end
      window_start += 1
    end
  end

  count
end

nums = [1,3,2,3,3]
k = 2
puts count_subarrays(nums, k) # 6ß